test_pii_data.py

```diff
@@ -29,61 +29,147 @@ def test_read_data(self):
+                         'Aggie Bred']
 
         data = read_data('sample_data.txt')
 
         self.assertEqual(data, expected_data)
+    def test_has_email(self):
+        test_data = Pii('My email is kaylahen2019@gmail.com')
+        self.assertEqual(test_data.has_email(), True)
+
+        test_data = Pii('My email is martin.complex@gmail.com')
+        self.assertEqual(test_data.has_email(), True)
+
+        test_data = Pii('My email is classof2023@aggies.ncat.edu')
+        self.assertEqual(test_data.has_email(), True)
+
+        test_data = Pii('My email is computerscience.com')
+        self.assertEqual(test_data.has_email(), None)
+
+        test_data = Pii('My email is engineering1gmail.com')
+        self.assertEqual(test_data.has_email(), None)
 
     def test_has_us_phone(self):
         # Test a valid US phone number
         test_data = Pii('My phone number is 970-555-1212')
         self.assertTrue(test_data.has_us_phone())
+        # Test a valid US phone number
+        test_data = Pii('My phone number is 9705551212')
+        self.assertTrue(test_data.has_us_phone())
 
         # Test a partial US phone number
         test_data = Pii('My number is 555-1212')
         self.assertFalse(test_data.has_us_phone())
 
-        # Test a phone number with incorrect delimiters
-        # TODO discuss changing requirements to support this
+        # Updated to allow for this entry
         test_data = Pii('My phone number is 970.555.1212')
-        self.assertFalse(test_data.has_us_phone())
+        self.assertTrue(test_data.has_us_phone())
 
     def test_has_email(self):
         test_data = Pii()
         self.assertEqual(test_data.has_email(), None)
 
     def test_has_ipv4(self):
-        test_data = Pii()
-        self.assertEqual(test_data.has_ipv4(), None)
+        #Successful test cases
+        test_data = Pii('222.33.100.12')
+        self.assertTrue(test_data.has_ipv4())
+
+        test_data = Pii('45.68.195.254')
+        self.assertTrue(test_data.has_ipv4())
+
+        # Test values are out of the range for allowed addresses
+        test_data = Pii('300.22.555.256')
+        self.assertFalse(test_data.has_ipv4())
+
+        # Test wrong format
+        test_data = Pii('145')  # incomplete address
+        self.assertFalse(test_data.has_ipv4())
+
 
     def test_has_ipv6(self):
-        test_data = Pii()
-        self.assertEqual(test_data.has_ipv6(), None)
+        # test a valid address
+        test_data = Pii('0045:Fa34:53d9:4d53:0020:0000:6491:8485')
+        self.assertTrue(test_data.has_ipv6())  
+
+        test_data = Pii('0000::5248:ee43::8789:1234:1200')
+        self.assertTrue(test_data.has_ipv6())  
+
+        # test an invalid address with to many digits in segment
+        test_data = Pii('00000:::::')
+        self.assertFalse(test_data.has_ipv6())        
+        
+        # test an invalid address w letter outside of bounds
+        self.assertFalse(test_data.has_ipv6())  
```

Comment:
This looks backwards and would actually just re-check the previous one since the assignment is after the has_ check
```
st_data = Pii('00000:::::')
```